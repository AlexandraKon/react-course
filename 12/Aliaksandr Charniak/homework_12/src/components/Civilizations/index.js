import { connect } from 'react-redux'

import Civilizations from './Civilizations'

const mapStateToProps = (state) => ({
  data: state.civilizations,
})

export default connect(mapStateToProps)(Civilizations)
