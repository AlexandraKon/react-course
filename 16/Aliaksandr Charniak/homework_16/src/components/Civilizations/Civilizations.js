import React, { useEffect } from 'react'
import { useDispatch } from 'react-redux'
import { Link } from 'react-router-dom'

import { getCorrectedArray } from '../../utils/utils'
import { ACTION_GET_CIVILIZATIONS_REQUESTED } from '../../redux/civilizations/actions'

export const Civilizations = (props) => {
  const { data } = props.data
  const { state } = props.location
  const arr = getCorrectedArray(data, 'civilizations')
  const dispatch = useDispatch()
  useEffect(() => {
    dispatch(ACTION_GET_CIVILIZATIONS_REQUESTED(state?.unitUrl))
  }, [dispatch, state?.unitUrl])
  return (
    <div>
      {arr &&
        arr.map((item) => (
          <ul key={item.id}>
            <li>Id: {item.id}</li>
            <li>Name: {item.name}</li>
            <li>Expansion: {item.expansion}</li>
            <li>Army type: {item.army_type}</li>
            <li>
              Unique unit:
              {item.unique_unit.map((it, i) => {
                const unitName = it.substring(it.lastIndexOf('/') + 1)
                return (
                  <ul key={i}>
                    <Link
                      to={{
                        pathname: `/civilizations/${item.id}/unique_unit/${unitName}`,
                        state: {
                          unitUrl: it,
                        },
                      }}
                    >
                      {unitName}
                    </Link>
                  </ul>
                )
              })}
            </li>
            <li>
              Unique tech:
              {item.unique_tech.map((it, i) => {
                const techName = it.substring(it.lastIndexOf('/') + 1)
                return (
                  <ul key={i}>
                    <Link
                      to={{
                        pathname: `/civilizations/${item.id}/unique_tech/${techName}`,
                        state: {
                          techUrl: it,
                        },
                      }}
                    >
                      {techName}
                    </Link>
                  </ul>
                )
              })}
            </li>
            <li>Team bonus: {item.team_bonus}</li>
            <li>
              Civilization bonus:
              {item.civilization_bonus.map((it, i) => (
                <ul key={i}>
                  <li>{it}</li>
                </ul>
              ))}
            </li>
          </ul>
        ))}
    </div>
  )
}

export default Civilizations
