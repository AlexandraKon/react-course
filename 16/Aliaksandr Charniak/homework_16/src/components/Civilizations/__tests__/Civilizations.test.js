import React from 'react'
import { configure, shallow } from 'enzyme'
import Adapter from '@wojtekmaj/enzyme-adapter-react-17'
import configureMockStore from 'redux-mock-store'

import { Provider } from 'react-redux'

import Civilizations from '..'

configure({ adapter: new Adapter() })

const mockStore = configureMockStore()
const store = mockStore({})

describe('<Civilizations /> mounting behaviour', () => {
  const props = {
    data: {
      civilizations: [
        {
          id: 1,
          name: 'Aztecs',
          expansion: 'The Conquerors',
          army_type: 'Infantry and Monk',
          unique_unit: ['https://age-of-empires-2-api.herokuapp.com/api/v1/unit/jaguar_warrior'],
          unique_tech: [
            'https://age-of-empires-2-api.herokuapp.com/api/v1/technology/garland_wars',
          ],
          team_bonus: 'Relics generate +33% gold',
          civilization_bonus: [
            'Villagers carry +5',
            'Military units created 15% faster',
            '+5 Monk hit points for each Monastery technology',
            'Loom free',
          ],
        },
        {
          id: 2,
          name: 'Britons',
          expansion: 'Age of Kings',
          army_type: 'Foot Archer',
          unique_unit: ['https://age-of-empires-2-api.herokuapp.com/api/v1/unit/longbowman'],
          unique_tech: ['https://age-of-empires-2-api.herokuapp.com/api/v1/technology/yeomen'],
          team_bonus: 'Archery Ranges work 20% faster',
          civilization_bonus: [
            'Town Centers cost -50% wood upon reaching the Castle Age',
            'Foot archers (excluding Skirmishers) have +1 range in Castle Age and +1 in Imperial Age (for +2 total)',
            'Shepherds work 25% faster',
          ],
        },
      ],
      units: 'https://age-of-empires-2-api.herokuapp.com/api/v1/units',
      structures: 'https://age-of-empires-2-api.herokuapp.com/api/v1/structures',
      technologies: 'https://age-of-empires-2-api.herokuapp.com/api/v1/technologies',
    },
    location: false,
  }

  let wrapper

  beforeEach(() => {
    wrapper = shallow(
      <Provider store={store}>
        <Civilizations {...props} />
      </Provider>
    )
  })

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('should render 1 <Civilizations /> component', () => {
    expect(wrapper).toHaveLength(1)
  })
})
