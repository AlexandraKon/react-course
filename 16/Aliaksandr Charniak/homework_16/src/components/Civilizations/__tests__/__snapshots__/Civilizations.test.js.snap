// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Civilizations /> mounting behaviour should match snapshot 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <Connect(Civilizations)
    data={
      Object {
        "civilizations": Array [
          Object {
            "army_type": "Infantry and Monk",
            "civilization_bonus": Array [
              "Villagers carry +5",
              "Military units created 15% faster",
              "+5 Monk hit points for each Monastery technology",
              "Loom free",
            ],
            "expansion": "The Conquerors",
            "id": 1,
            "name": "Aztecs",
            "team_bonus": "Relics generate +33% gold",
            "unique_tech": Array [
              "https://age-of-empires-2-api.herokuapp.com/api/v1/technology/garland_wars",
            ],
            "unique_unit": Array [
              "https://age-of-empires-2-api.herokuapp.com/api/v1/unit/jaguar_warrior",
            ],
          },
          Object {
            "army_type": "Foot Archer",
            "civilization_bonus": Array [
              "Town Centers cost -50% wood upon reaching the Castle Age",
              "Foot archers (excluding Skirmishers) have +1 range in Castle Age and +1 in Imperial Age (for +2 total)",
              "Shepherds work 25% faster",
            ],
            "expansion": "Age of Kings",
            "id": 2,
            "name": "Britons",
            "team_bonus": "Archery Ranges work 20% faster",
            "unique_tech": Array [
              "https://age-of-empires-2-api.herokuapp.com/api/v1/technology/yeomen",
            ],
            "unique_unit": Array [
              "https://age-of-empires-2-api.herokuapp.com/api/v1/unit/longbowman",
            ],
          },
        ],
        "structures": "https://age-of-empires-2-api.herokuapp.com/api/v1/structures",
        "technologies": "https://age-of-empires-2-api.herokuapp.com/api/v1/technologies",
        "units": "https://age-of-empires-2-api.herokuapp.com/api/v1/units",
      }
    }
    location={false}
  />
</ContextProvider>
`;
